src/CoreRCON/RCON.cs

```diff
@@ -367,39 +369,48 @@ public async Task<string> SendCommandAsync(string command, TimeSpan? overrideTim
         private void RCONPacketReceived(RCONPacket packet)
         {
             _logger?.LogTrace("RCON packet received: {}", packet.Id);
+            
+            TaskCompletionSource<string> taskSource;
             // Call pending result and remove from map
-            if (_pendingCommands.TryGetValue(packet.Id, out TaskCompletionSource<string> taskSource))
+            if (!_pendingCommands.TryGetValue(packet.Id, out taskSource))
             {
-                if (_multiPacket)
+                _logger?.LogWarning("Received packet with no matching command id: {} body: {}", packet.Id, packet.Body);
+                // The server did not respect our id
+                if (!strictCommandPacketIdMatching && packet.Id == 0)
                 {
-                    //Read any previous messages 
-                    _incomingBuffer.TryGetValue(packet.Id, out string body);
-
-                    if (packet.Body == Constants.MULTI_PACKET_END_RESPONSE)
-                    {
-                        //Avoid yielding
-                        taskSource.SetResult(body ?? string.Empty);
-                        _pendingCommands.TryRemove(packet.Id, out _);
-                        _incomingBuffer.Remove(packet.Id);
-                    }
-                    else
-                    {
-                        //Append to previous messages
-                        _incomingBuffer[packet.Id] = body + packet.Body;
-                    }
+                    // Get the most recently sent command
+                    taskSource = _pendingCommands
+                        .OrderBy(cmd => cmd.Key)
+                        .Select(cmd => cmd.Value)
+                        .FirstOrDefault();
                 }
-                else
+            }
+
+            if (_multiPacket)
+            {
+                //Read any previous messages 
+                _incomingBuffer.TryGetValue(packet.Id, out string body);
+
+                if (packet.Body == Constants.MULTI_PACKET_END_RESPONSE)
                 {
                     //Avoid yielding
-                    taskSource.SetResult(packet.Body);
+                    taskSource.SetResult(body ?? string.Empty);
                     _pendingCommands.TryRemove(packet.Id, out _);
+                    _incomingBuffer.Remove(packet.Id);
+                }
+                else
+                {
+                    //Append to previous messages
+                    _incomingBuffer[packet.Id] = body + packet.Body;
                 }
             }
             else
             {
-                _logger?.LogWarning("Received packet with no matching command id: {} body: {}", packet.Id, packet.Body);
+                //Avoid yielding
+                taskSource.SetResult(packet.Body);
```

Comment:
taskSource can be null here if `strictCommandPacketIdMatching == false`.
Add a null check.

Reply:
Added a gaurd clause, should prevent this